### ISSUE TITLE
Implement plugin system

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#1106]
This is a roll up issue for implementing the plugin system. Individual tasks:
- [x] Make `mergeConfigs()` merge plugin entries  (#1111)
- [x] Add `--plugin` option to CLI (#1112)
- [x] Update `CLIEngine` to load plugins from command line (#1113)
- [x] Add `rules.import()` to import rules from another object (#1114)
- [x] Update `CLIEngine` to load plugins from configs (#1115)
- [x] Update comment parsing to read rules with slashes (#1116)
- [x] Plugin documentation (#1117)
### Plugins
- Each plugin is an npm module with a name in the format of `eslint-plugin-<plugin-name>`, such as `eslint-plugin-jquery`.
  - This ensures uniqueness for all plugins published on the npm registry.
  - It also allows users to easily search for and find plugins.
  - The naming patterns also leaves the door open for creating other tools to easily find plugins.
- The module must export an object with a `rules` property. This `rules` property should be an object containing a key-value mapping of rule ID to rule (what is in our rule files today).
  - The rule ID does not have to follow any naming convention (so it can just be `dollar-sign`, for instance).
  - Leaves open the possibility exporting things like environments in the future
### Loading plugins

In order to use a plugin, you must first install it using npm. Once installed, you can specify which plugins to use either on the command line using the `--plugin` option or in a `.eslintrc` file. In both cases, you can omit "eslint-plugin-" from the name and it will still work:

```
# Loads eslint-plugin-jquery
eslint --plugin jquery .
```

In a configuration file:

``` yaml
env:
  browser: true
plugins:
  - jquery   # loads eslint-plugin-jquery
rules:
  semi: 2
```
### Configuring rules

In order to configure plugin rules, use the plugin name (minus the "eslint-plugin-") plus the rule ID  separated with a slash (such as `jquery/dollar-sign`). The slash disambiguates plugin rules from normal rules while allowing us to continue using the same logic as we've been using. This can be used both in configuration files and inline. Here's a configuration file:

``` yaml
env:
  browser: true
plugins:
  - jquery   # loads eslint-plugin-jquery
rules:
  semi: 2
  jquery/dollar-sign: 2    # turn on plugin rule
```

To configure inline:

``` js
/*eslint semi: 2, jquery/dollar-sign: 2*/

/*eslint-disable jquery/dollar-sign*/
```
### Answering Open Questions

> What to do about naming collisions for rules? Should all rules automatically be prefixed with the node module name to avoid collisions? Should we leave it up to rule developers to create prefixed rules on their own?

There can be no naming collisions. All rules are automatically prefixed with their plugin name as a namespace. If an attempt is made to add a rule with the same namespace and ID as an already-registered one, then ESLint should output an error and exit immediately. 

> Should you be able to specify rules in a plugin to be on by default? (Have their own eslint.json equivalent?)

Not now, but will revisit in the future.

> If a specified plugin is missing, is that a fatal error or a warning?

Fatal error, ESLint should output an error and exit immediately.
### Out of Scope

There are several other issues that emerged and that I'm purposely deferring until a later point in time:
- What about loading plugins from a local file rather than an npm package?
- Can plugins specify some rules to be on by default?
- Can plugins define new environments?
### Conclusion

I think that this solves all of the high-level problems we're looking to solve without painting us into a corner. We can always relax restrictions in the future if necessary, whereas it's harder to implement new restrictions where there previously weren't any.
